### DISCUSSION TITLE
Dúvida com relação a camada Command da aplicação

### DISCUSSION THREAD

**Discussion by RonaldoBueno16** [balta-io/forum#399] [Tira Dúvidas] [answered]
Curso: Todo App: API com ASP.NET Core, CQRS e EF Core (13%) 
Aula: Estrutura dos Commands

Balta, boa tarde!

Mudando um pouco o cenário da API, considerando que é uma API de Carrinho de Compras.

Na API, seria viável ter um Comando para cada ação da minha aplicação? Exemplo:

- Adicionar produto no carrinho
- Recuperar carrinho
- Remover produto no carrinho
- Limpar carrinho

Isso deixaria a aplicação mais complexa, mas é uma boa prática? Levando em consideração o escalonamento, testes e manutenção da aplicação?

Obrigado!

**Comment by henriqueweege** [comment, marked as answer]
Oi @RonaldoBueno16 , tudo bem?

Sei que você chamou pelo Balta, mas acho que posso te ajudar. Te peço desculpas desde já por me intrometer aqui.

Quando se trata de CQRS é importante se ter em mente que todas as ações que tem como propósito escrita de dados são Commands, e todas as ações que dizem respeito a leitura de dados são Queries. Essa distinção é importante pois, acredito que no caso que você citou, uma estrutura de CRUD já daria conta das operações.

O que eu acredito que seja uma saída interessante nesses casos em que há um olhar para a escalabilidade, é seguir POO e ter um Handler abstrato para CRUD, com métodos possíveis de serem sobrescritos. Dessa forma, caso exista alguma regra de negócio envolvida em alguma dessas operações, ainda é possível implementar (e de quebra ainda segue o Open-Close Principle do SOLID), e caso não haja, já está feita a implementação. Caso queira algum exemplo disso, fiz o melhor que pude nesse sentido no projeto que eu codei no final do curso 2813, e está na pasta "Exercício" [desse repositório](https://github.com/henriqueweege/2813). O handler em questão está em Domain/RoomBooking.BuniessLogic/Handlers/Base/CRUDHandler.

Quanto ao cenário que você  citou, eu modelaria a aplicação um pouco diferente. Não tenho certeza se é a melhor saída pois nunca codei na vida real uma aplicação que necessitasse de um carrinho de compras, porém, me parece que seria melhor enviar ao back-end uma ordem de compra já completa, com todos os produtos selecionados.

Enfim, espero ter ajudado, peço desculpas antecipadas por qualquer erro, e qualquer coisa que eu possa ajudar é só me dar um toque. 
Abração!

**Comment by RonaldoBueno16** [comment]
Henrique, tudo joia e você?

Quando você diz sobre ter um Handler abstrato para o CRUD, você se refere a camada de dominio da aplicação?

Pois no que eu citei, eu penso na seguinte estrutura para um carrinho de compras:

CarrinhoDeCompras/CarrinhoDeCompras.Api
CarrinhoDeCompras/CarrinhoDeCompras.Domain

- Entities
Product -> Aqui ficará as propriedades do produto
Cart -> Aqui ficará a lista de Produto
- Commands
AddCartProductCommand
RemoveCartProductCommand
GetCartCommand
ClearCartCommand
- Queries
GetCartQuerie
- Repositories
CartRepository
ProductRepository
- Services
CartService
ProduceService

Desconsiderando as outras camadas da aplicação, tipo a de Infra, para ficar a burocracia com autenticação com banco de dados.
Eu pensei nessa estrutura para um carrinho de compras

- Por que não enviar a ordem de compra já pronta para o back?
Pois é um pré-requisito o usuário ter esse carrinho de compras salvo no banco de dados, para caso ele abandone o carrinho, enviar um e-mail para ele informando que o carrinho foi abandonado.

O que você acha?

**Reply by henriqueweege** [reply]
Oi, Ronaldo!
Sim, me refiro a camada de domínio da aplicação. Tudo depende da modelagem que a aplicação vai ter, porém, as operações que você citou cabem dentro de um CRUD. 

Como eu modelaria a camada de domínio:

Projeto.Entities

     Entities
          Product -> Aqui ficará as propriedades do produto
          Order -> Aqui ficará a lista de Produto daquela compra

Projeto.BusinessLogic

     Commands (Lembrando que Commands em CQRS são responsáveis apenas por escrita de dados, não por leitura de dados)
          CreateProductCommand
          UpdateProductCommand
          DeleteProductCommand

     Queries (Lembrando que Queries em CQRS são responsáveis apenas por leitura de dados, não por escrita de dados)
          GetProductQuery
          GetProductByIdQuery
          GetOrderQuery
          GetOrderByIdQuery

     Handlers
          CRUDHandler -> Handler genérico para lidar com CRUD sem a necessidade de criar método por método 
                                      de todas as entidades,           
                                      porém, abstrato em com possibilidade de sobrescrita dos métodos para o caso de
                                       necessidade de implementação própria da entidade.
          ProductHandler -> herda CRUDHandler, implementa as lógicas necessárias.
          OrderHandler -> herda de CRUDHandler, implementa as lógicas necessárias.

Como nesse caso os repositórios não teriam nenhuma regra de negócio, eu os abstrairia para a camada de Infrastructure junto com todos os acesso a partes fora da API.

Por que não enviar a ordem de compra já pronta para o back?
Bom, isso depende de cada projeto e tudo tem tradeoff. Eu não sou dono da verdade, e provavelmente outras pessoas modelariam a aplicação de uma maneira até mais interessante do que a que eu estou sugerindo. Dito isso, eu particularmente não acho que seja uma boa ideia salvar o carrinho de compras,  pois me parece que da maneira proposta poderia gerar alguns problemas. Dependendo de como for implementado, os itens que estão no carrinho não poderiam estar disponíveis para outra pessoa comprar, e ai por querer garantir a compra do cliente X, você pode perder a compra do cliente Y por uma indisponibilidade que não é real. Quanto tempo essa compra fica garantida? E no caso de alguém fazer um carrinho de compras com todos os items do teu estoque e não finalizar? Há a necessidade de garantir a compra? Caso não haja, qual o motivo de onerar a infra da empresa com todos esses dados?
Além disso, você não tem garantia nenhuma que essa compra vai ser realizada. Imagina o cenário de uma loja grande qualquer, as centenas de milhares de compras que ocorrem nessas lojas e a quantidade de ocasiões onde se cria um carrinho de compra e não se cria a ordem de compra. Essa quantidade de dados vai ficar persistida no banco? Isso pode gerar milhões de registros que nunca vão se converter em compra, portanto, só onerando a infra da empresa, o que aumenta os custos da operação. 
Enfim, o que eu faria: o cliente está com um carrinho com itens e encerrou a sessão dele sem terminar a compra? O front envia para o back os ids dos itens do carrinho, e o back dispara um e-mail para o cliente, sem de fato persistir nada no banco ou atribuir produto a compra não realizada. Ainda existe a possibilidade de deixar esse carrinho salvo num LocalStorage pelo front, por exemplo, e quando o cliente voltar a página, pode requerer ao back para que valide a Ordem a ser gerada para ter certeza que não há indisponibilidade dos produtos.  
As possibilidades são muitas, e as vezes as pessoas quererem o que quererem.

Abração!!
